File: Anurag_19GCEBCS079/Python/day4.py
# Link For Problem: https://leetcode.com/problems/max-number-of-k-sum-pairs/

from collections import Counter
from math import floor


class Solution:

    """
      Put the array elements
      in a map and keep the frequency
      count.

      Check if (k-i) exists
      in the map or not.

      TC: O(n) + O(n) = 2O(n) => O(n)
      SC: O(n)

    """

    def maxOperations(self, nums: list[int], k: int) -> int:
        map = {}

        for i in nums:
            if i >= k:
                continue

            if i in map:
                map[i] += 1
            else:
                map[i] = 1

        count: int = 0

        for i in nums:
            if i >= k:
                continue

            diff: int = k-i

            if diff == i and diff in map:
                count += floor(map[diff]/2)
                map.pop(diff)

            elif i in map and diff in map:
                count += min(map[i], map[diff])
                map.pop(i)
                map.pop(diff)

        return count

    def another_solution(self, nums: list[int], k: int) -> int:
        map = Counter(nums)

        count: int = 0

        for i in nums:
            if i >= k:
                continue

            diff: int = k-i

            if diff == i and diff in map:
                count += floor(map[diff]/2)
                map.pop(diff)

            elif i in map and diff in map:
                count += min(map[i], map[diff])
                map.pop(i)
                map.pop(diff)

        return count

    """
     Simply Sort the
     Array.
      
     Use Two Pointer Approach
     to check if the sum is
     equal to k.
      
     TC: O(nlogn) + O(n) => O(nlogn)
     SC: O(1)

    """

    def two_pointer(self, nums: list[int], k: int) -> int:
        nums.sort()

        i, j = 0, len(nums)-1
        count: int = 0

        while(i < j):
            sum: int = nums[i] + nums[j]

            if sum == k:
                count += 1
                i += 1
                j -= 1

            elif sum < k:
                i += 1

            else:
                j -= 1

        return count

File: Anurag_19_CSE/Java/day4.java
//Link For Problem: https://leetcode.com/problems/max-number-of-k-sum-pairs/

import java.util.Arrays;
import java.util.HashMap;

public class day4 {

    /*
     * Put the array elements
     * in a map and keep the frequency
     * count.
     * 
     * Check if (k-i) exists
     * in the map or not.
     * 
     * TC: O(n) + O(n) = 2O(n) => O(n)
     * SC: O(n)
     */

    public int maxOperations(int[] nums, int k) {
        HashMap<Integer, Integer> hm = new HashMap<>();

        for (int i : nums) {
            if (i >= k)
                continue;

            hm.put(i, hm.getOrDefault(i, 0) + 1);
        }

        int count = 0;
        for (int i : nums) {
            if (i >= k)
                continue;

            int diff = k - i;

            if (diff == i && hm.containsKey(diff)) {
                count += Math.floor(hm.get(diff) / 2);
                hm.remove(diff);
            }

            else if (hm.containsKey(diff) && hm.containsKey(i)) {
                count += Math.min(hm.get(i), hm.get(diff));
                hm.remove(diff);
                hm.remove(i);
            }
        }

        return count;
    }

    /*
     * Simply Sort the
     * Array.
     * 
     * Use Two Pointer Approach
     * to check if the sum is
     * equal to k.
     * 
     * TC: O(nlogn) + O(n) => O(nlogn)
     * SC: O(1)
     */

    public int twoPointer(int nums[], int k) {
        Arrays.sort(nums);

        int i = 0, j = nums.length - 1;
        int count = 0;

        while (i < j) {
            int sum = nums[i] + nums[j];

            if (sum == k) {
                count += 1;
                i += 1;
                j -= 1;
            }

            else if (sum < k)
                i += 1;

            else
                j -= 1;
        }

        return count;
    }
}
